Added support for SFTTrainer checkpoint models and adapter models containing some non-LoRA weights

Victoran0 opened the pull request:
The previous code triggers an unexpected name error and calls sys.exit(1) (lines 350-351 current version) even if a single weight in the lora_model is not a lora_A, lora_B, or base layer weight. 
This edit collects the names of all LoRA weights in the model before the for loop in line 341 (current version). And in line 350 (edit version), the subsequent operations are performed only on the LoRA and base layer weights, ignoring any non-LoRA weights in the lora_model. 
Hopefully, this helps by allowing the script to extract LoRA weights and convert LoRA to GGUF for adapter models containing one or more non-LoRA weights.



- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [ ] Medium
  - [x] High

Victoran0 committed:
Added support for SFTTrainer checkpoint models and adapter models containing some non-LoRA weights

The previous code triggers an unexpected name error and calls sys.exit(1) (lines 350-351 current version) even if a single weight in the lora_model is not a lora_A, lora_B, or base layer weight.
This edit collects the names of all LoRA weights in the model before the for loop in line 341 (current version). And in line 350 (edit version), the subsequent operations are performed only on the LoRA and base layer weights, ignoring any non-LoRA weights in the lora_model.
Hopefully, this helps by allowing the script to extract LoRA weights and convert LoRA to GGUF for adapters containing one or more non-LoRA weights.

FirstTimeEZ commented on convert_lora_to_gguf.py:
```diff
@@ -338,28 +338,39 @@ def generate_extra_tensors(self) -> Iterable[tuple[str, Tensor]]:
-                            tensor_map[base_name] = PartialLoraTensor(A=tensor)
+
+                    # Check for only LoRA finetuned weights and base layer weights
+                    if (name in lora_model_items_with_lora_tensors) or (".base_layer.weight" in name):
```
why don't you check `"lora_"` the same way you checked `".base_layer.weight"`

change line 350

```py
if (name in lora_model_items_with_lora_tensors) or (".base_layer.weight" in name):
```

to the following

```py
if ("lora_" in name) or (".base_layer.weight" in name):
```

then remove line 345

```
lora_model_items_with_lora_tensors = [name for name in lora_model_items_name if 'lora_' in name]
```

it seems like `lora_model_items_with_lora_tensors` exists for no reason, please explain.

FirstTimeEZ commented on convert_lora_to_gguf.py:
```diff
@@ -338,28 +338,39 @@ def generate_extra_tensors(self) -> Iterable[tuple[str, Tensor]]:
             def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
```
it seems like `lora_model_items_with_lora_tensors` and `lora_model_items_name` exist for no reason

You should change `350` and remove `344` and `345`:

```py
def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
    tensor_map: dict[str, PartialLoraTensor] = {}

    for name, tensor in lora_model.items():
        if ("lora_" in name) or (".base_layer.weight" in name):
            if self.lazy:
                tensor = LazyTorchTensor.from_eager(tensor)
            base_name = get_base_tensor_name(name)
            is_lora_a = ".lora_A.weight" in name
            is_lora_b = ".lora_B.weight" in name
            if not is_lora_a and not is_lora_b:
                if ".base_layer.weight" in name:
                    continue

            if base_name in tensor_map:
                if is_lora_a:
                    tensor_map[base_name].A = tensor
                else:
                    tensor_map[base_name].B = tensor
            else:
                if is_lora_a:
                    tensor_map[base_name] = PartialLoraTensor(A=tensor)
                else:
                    tensor_map[base_name] = PartialLoraTensor(B=tensor)
        else:
            pass                

    for name, tensor in tensor_map.items():
        assert tensor.A is not None
        assert tensor.B is not None
        yield (name, cast(torch.Tensor, LoraTorchTensor(tensor.A, tensor.B)))
```

This is functionally the same and will be more performant.

please explain the reasoning behind the added complexity or if this is suitable.

FirstTimeEZ replied:
it seems like `lora_model_items_with_lora_tensors` and `lora_model_items_name` exist for no reason

You should change `350` and remove `344` and `345`:

```py
def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
    tensor_map: dict[str, PartialLoraTensor] = {}

    for name, tensor in lora_model.items():
        if ("lora_" in name) or (".base_layer.weight" in name):
            if self.lazy:
                tensor = LazyTorchTensor.from_eager(tensor)
            base_name = get_base_tensor_name(name)
            is_lora_a = ".lora_A.weight" in name
            is_lora_b = ".lora_B.weight" in name
            if not is_lora_a and not is_lora_b:
                if ".base_layer.weight" in name:
                    continue

            if base_name in tensor_map:
                if is_lora_a:
                    tensor_map[base_name].A = tensor
                else:
                    tensor_map[base_name].B = tensor
            else:
                if is_lora_a:
                    tensor_map[base_name] = PartialLoraTensor(A=tensor)
                else:
                    tensor_map[base_name] = PartialLoraTensor(B=tensor)
        else:
            pass                

    for name, tensor in tensor_map.items():
        assert tensor.A is not None
        assert tensor.B is not None
        yield (name, cast(torch.Tensor, LoraTorchTensor(tensor.A, tensor.B)))
```

This is functionally the same and will be more performant.

please explain the reasoning behind the added complexity or if this is suitable.

FirstTimeEZ commented on convert_lora_to_gguf.py:
```diff
@@ -338,28 +338,39 @@ def generate_extra_tensors(self) -> Iterable[tuple[str, Tensor]]:
             def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
                 tensor_map: dict[str, PartialLoraTensor] = {}
 
+                # The following edits will enable conversion for: SFTTrainer checkpoint adapter models and other adapter models that contain weights besides LoRA weights
```
it seems like `lora_model_items_with_lora_tensors` and `lora_model_items_name` exist for no reason

You could check `("lora_" in name)` and `pass` when there is no match instead:

```py
def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
    tensor_map: dict[str, PartialLoraTensor] = {}

    for name, tensor in lora_model.items():
        if ("lora_" in name) or (".base_layer.weight" in name):
            if self.lazy:
                tensor = LazyTorchTensor.from_eager(tensor)
            base_name = get_base_tensor_name(name)
            is_lora_a = ".lora_A.weight" in name
            is_lora_b = ".lora_B.weight" in name
            if not is_lora_a and not is_lora_b:
                if ".base_layer.weight" in name:
                    continue

            if base_name in tensor_map:
                if is_lora_a:
                    tensor_map[base_name].A = tensor
                else:
                    tensor_map[base_name].B = tensor
            else:
                if is_lora_a:
                    tensor_map[base_name] = PartialLoraTensor(A=tensor)
                else:
                    tensor_map[base_name] = PartialLoraTensor(B=tensor)
        else:
            pass                

    for name, tensor in tensor_map.items():
        assert tensor.A is not None
        assert tensor.B is not None
        yield (name, cast(torch.Tensor, LoraTorchTensor(tensor.A, tensor.B)))
```

This is functionally the same and will be more performant.

please explain the reasoning behind the added complexity or whether this is suitable.

FirstTimeEZ reviewed (changes requested):
You can remove `lora_model_items_with_lora_tensors` and `lora_model_items_name`

What you have done is the same as checking `("lora_" in name)` on each item in `lora_model.items()`

Everything in `lora_model.items()` needs to be checked anyway so you can just `pass` on things that don't match.

Here is a full example:

```py
def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
    tensor_map: dict[str, PartialLoraTensor] = {}

    for name, tensor in lora_model.items():
        if ("lora_" in name) or (".base_layer.weight" in name):
            if self.lazy:
                tensor = LazyTorchTensor.from_eager(tensor)
            base_name = get_base_tensor_name(name)
            is_lora_a = ".lora_A.weight" in name
            is_lora_b = ".lora_B.weight" in name
            if not is_lora_a and not is_lora_b:
                if ".base_layer.weight" in name:
                    continue

            if base_name in tensor_map:
                if is_lora_a:
                    tensor_map[base_name].A = tensor
                else:
                    tensor_map[base_name].B = tensor
            else:
                if is_lora_a:
                    tensor_map[base_name] = PartialLoraTensor(A=tensor)
                else:
                    tensor_map[base_name] = PartialLoraTensor(B=tensor)
        else:
            pass                

    for name, tensor in tensor_map.items():
        assert tensor.A is not None
        assert tensor.B is not None
        yield (name, cast(torch.Tensor, LoraTorchTensor(tensor.A, tensor.B)))
```

This is functionally the same and will be more performant.

please explain the reasoning behind the added complexity or whether this is suitable.

Victoran0 committed:
Added support for SFTTrainer checkpoint models and adapter models containing one or more non-LoRA weights

My initial commit was more like a brute force.
The edits suggested by @FirstTimeEZ reduces the complexity.

Victoran0 reviewed (commented):
My initial commit was literally the Brute Force solution I implemented when I encountered the "Unexpected name '{name}': Not a lora_A or lora_B tensor" error while trying to convert a SFTTrainer-Checkpoint-lora-adapter-model to gguf.
The edits proposed by @FirstTimeEZ reduces the complexity of the code, making it more performant.

FirstTimeEZ reviewed (commented):
This looks fine to me now, but I don't know the reasoning behind `sys.exit(1)` being called in the original.

Victoran0 commented:
> This looks fine to me now, but I don't know the reasoning behind `sys.exit(1)` being called in the original.

Exactly, and the for loop checks for every weight in the lora_model and then `sys.exit(1)` if the weight is not a Lora weight.
The original will only work if all Lora weights have been extracted from the lora_model or every layer in the base_model was fine tuned.

ngxson reviewed (changes requested):
It make no sense to remove this error message.

We show the error because we never want to produce a half-working lora gguf. In your case, there are non-lora tensors in the file. This may hint that your `safetensors` file contains unknown data that is not handled by llama.cpp

It's better to know exactly what are the non-lora tensors in your case, rather than ignore them.

Please at least put a full output log here.

Victoran0 commented:
INFO:lora-to-gguf:Loading base model: 392a143b624368100f77a3eafaa4a2468ba50a72
INFO:gguf.gguf_writer:gguf: This GGUF file is for Little Endian only
INFO:lora-to-gguf:Exporting model...
ERROR:lora-to-gguf:Unexpected name 'base_model.model.lm_head.weight': Not a lora_A or lora_B tensor

Above is the logged output error.
My LoRA Adapter is a fine tuned meta-llama-3.2-3B-instruct. 
The target modules for my LoraConfig are `['up_proj', 'k_proj', 'v_proj', 'down_proj', 'q_proj', 'o_proj', 'gate_proj']`..
As you can see from the log above, the `lm_head` weight caused the `sys.exit(1)` code to run, because the `lm_head` is not part of the target modules that was fine tuned.
The script is limited to LoRA adapters with every layer fine-tuned.

ngxson commented:
@Victoran0 I'm checking with TRL team at hugging face to know why. In the meantime, could you share the adapter or the python code that you used to fine tune the model?

Victoran0 commented:
The Adapter: https://huggingface.co/Victorano/llama-3.2-3B-it-Procurtech-Assistant
The Python Code: https://colab.research.google.com/drive/1OC5puMjnWi4wGpOvE-pqr2Y4BqtiiP1f?usp=sharing
@ngxson Here you go!

ngxson commented:
Ok so the problem is that you used `setup_chat_format` which adds more tokens to the tokenizer. That's why it change the embeddings and lm_head, and this change is currently not supported by llama.cpp

What I suggest is to remove the `setup_chat_format`, because your base model `Llama-3.2-3B-Instruct` already have a chat template, so no need to setup.

The python code will also need to be updated to report the correct error message.

Victoran0 commented:
Okay, I will fix that, retrain and let you know the result

Victoran0 commented:
@ngxson , removing `setup_chat_format` resolved the error. The lora adapter's size is now less than 100mb, excellent!
The fix for #9948  is an optimal solution.
I guess I can close this now.
